services/motion/builtin/state/state.go

```diff
@@ -0,0 +1,647 @@
+	logger     golog.Logger
+	// mu protects the componentStateByComponent
+	mu                        sync.RWMutex
+	componentStateByComponent componentStateByComponent
+}
+
+// ExecutionReq is the request type for StartExecution.
+type ExecutionReq struct {
+	ComponentName      resource.Name
+	Destination        *geo.Point
+	Heading            float64
+	MovementSensorName resource.Name
+	Obstacles          []*spatialmath.GeoObstacle
+	MotionCfg          *motion.MotionConfiguration
+	Extra              map[string]interface{}
+}
+
+// PlanHistoryReq describes a request to PlansWithStatuses.
+type PlanHistoryReq struct {
+	ComponentName resource.Name
+	LastPlanOnly  bool
+	ExecutionID   motion.ExecutionID
+}
+
+// ListPlanStatusesReq describes a request to ListPlanStatuses.
+type ListPlanStatusesReq struct {
+	OnlyActivePlans bool
+}
+
+// NewState creates a new state.
+func NewState(logger golog.Logger) *State {
+	cancelCtx, cancelFunc := context.WithCancel(context.Background())
+	s := State{
+		cancelCtx:                 cancelCtx,
+		cancelFunc:                cancelFunc,
+		waitGroup:                 &sync.WaitGroup{},
+		componentStateByComponent: make(componentStateByComponent),
+		logger:                    logger,
+	}
+	return &s
+}
+
+// NewExecutionReq creates an ExecutionReq from a motion.MoveOnGlobeReq.
+func NewExecutionReq(req motion.MoveOnGlobeReq) ExecutionReq {
+	return ExecutionReq{
+		ComponentName:      req.ComponentName,
+		Destination:        req.Destination,
+		Heading:            req.Heading,
+		MovementSensorName: req.MovementSensorName,
+		Obstacles:          req.Obstacles,
+		MotionCfg:          req.MotionCfg,
+		Extra:              req.Extra,
+	}
+}
+
+// NewPlanHistoryReq creates a PlanHistoryReq from a motion.PlanHistoryReq.
+func NewPlanHistoryReq(req motion.PlanHistoryReq) PlanHistoryReq {
+	return PlanHistoryReq{
+		ExecutionID:   req.ExecutionID,
+		ComponentName: req.ComponentName,
+		LastPlanOnly:  req.LastPlanOnly,
+	}
+}
+
+// NewListPlanStatusesReq creates a ListPlanStatusesReq from a motion.ListPlanStatusesReq.
+func NewListPlanStatusesReq(req motion.ListPlanStatusesReq) ListPlanStatusesReq {
+	return ListPlanStatusesReq{
+		OnlyActivePlans: req.OnlyActivePlans,
+	}
+}
+
+// StartExecution creates a new execution from a state.
+func (s *State) StartExecution(req ExecutionReq, tc *TestConfig) (motion.ExecutionID, error) {
+	if err := s.validateNoActiveExecutionID(req.ComponentName); err != nil {
+		return uuid.Nil, err
+	}
+
+	// the state being cancelled should cause all executions derived from that state to also be cancelled
+	cancelCtx, cancelFunc := context.WithCancel(s.cancelCtx)
+	e := execution{
+		id:            uuid.New(),
+		state:         s,
+		componentName: req.ComponentName,
+		cancelCtx:     cancelCtx,
+		cancelFunc:    cancelFunc,
+		waitGroup:     &sync.WaitGroup{},
+		logger:        s.logger,
+		testConfig:    addExtra(tc, req.Extra),
+	}
+
+	p, err := e.newPlan()
```

Comment:
[opt] I think this will also need the destination, heading, etc. Not sure if you are planning to add that in this ticket, or a later ticket.

Reply:
yep, I plan on adding that in this ticket when we implement the actual business logic of MoveOnGlobe: https://viam.atlassian.net/browse/RSDK-4814

Reply:
sounds good